### DISCUSSION TITLE
Getting started with Nomadnetwork

### DISCUSSION THREAD

**Discussion by Dualfuel671** [markqvist/Reticulum#166] [Nomad Network]
Hi Mark, and all,
I have started learning Nomadnetwork.
I started with a Toshiba 50 laptop with Ubuntu 20.4.
I checked the laptop for python by opening a terminal and simply typing "pip". Oops command not found but if I typed in install pip, I would get it...so I did.
I am trying to be painstaking here because it is always the details that prevent me from succeeding not the big picture.
My install command was the same as in the getting started fast guide.
Yes! I did have to reboot the laptop!!!!

**Comment by Dualfuel671** [comment]
/home/dualfuel/Pictures/Screenshot from 2022-11-28 07-34-07.png

I tried the Frankfurt test net and failed. I am too new at this to know why yet.
I typed in the Dublin address and got a connection refused.
When the guide says to C-u and type in the address, I assumed you meant the 16bit address correct?
I may have found a bug....when I am in the conversations screen, and hit C-e to see peer info, I get kicked out of the program. I don't know where to leave those comments yet.

finally, Mark, do you have a copy of the entries I left on the old forum? I left detailed instructions for how to deploy both MFJ and Kantronics TNCs using TNCattach. It would save me an incredible amount of backtracking if I could read those instructions.

**Comment by markqvist** [comment]
Hi @Dualfuel671!
Thanks for documenting your progress and providing these notes.

Aha, that is my fault. It seems like the connection details in the internal guide were outdated. You can use the following if you want:

```
[[RNS Testnet Zurich]]
  type = TCPClientInterface
  interface_enabled = yes
  outgoing = True
  target_host = zurich.connect.reticulum.network
  target_port = 4242
```

Or alternatively one of the nodes listed here: https://reticulum.network/connect.html

I'll send you a link to the old forum, then you should be able to dig out those instructions :)

> I typed in the Dublin address and got a connection refused.

This is most likely since Reticulum could not find a path to the actual node, seeing that the testnet connection was not working.

> I may have found a bug....when I am in the conversations screen, and hit C-e to see peer info, I get kicked out of the program. I don't know where to leave those comments yet.

Thanks, souds like it. I will investiage it!

**Comment by Dualfuel671** [comment]
Its important to know that the reticulum config file can be found by typing cd /home/"your computer"/.reticulum
if you are like me, and not too savy on Linux...I forget to use that "." to reveal that hidden file.
All this in a terminal!
Ok, I think I need to learn how to use cut and paste within Nomad network. Its pretty tedious writing in the 16bit addresses.
I need to research whether Esp32 or Esp8266 can use Reticulum.
I added the Zurich testnet but I think I would like to see a screen shot of what the Nomad net looks like when it is connected properly.
its all good learning tho... :)

**Reply by Dualfuel671** [reply]
when in Nomad network, and clicking on "[ Config ]" if you can't move the cursor, or write to the editor, just do a Control -o and save the file. After that, the editor will work.

**Comment by Dualfuel671** [comment]
"I would like to see a screen shot of what the Nomad net looks like when it is connected properly." I wrote this earlier and then scrolled down the Nomadnet github page and saw Mark's Screenshots. My screen doesn't look like those. I don't mean the Micron business either, I mean it doesn't have the same format. I will continue poking around and maybe make it match.

**Comment by Dualfuel671** [comment]
About installing Nomadnet on a Raspberry Pi 3B+....
I updated the Raspberry Pi OS
I sudo python -m pip install --upgrade pip
I sudo pip3 install nomadnet
after a while....I got a fail....because I didn't have Rust installed so...
I sudo apt-get install build essential libssl-dev     python3-dev cargo      //this from the Rust website
Still failed to build cryptography. failed with code 101...
the error was too long to write down here....I am going to reboot the pi and try to install nomadnet again.

After trying everything I can think of, I am stranded at this point in the installation....
"Building wheel for cryptography (pyproject.toml) ... /"
it seems like it goes into this subroutine and never returns, I have to Control-C to get out.

**Comment by Dualfuel671** [comment]
I used a different SD card with a 32bit Raspberry Pi OS installed. 
I sudo apt-get upgrade
I pip install pip --upgrade
(upgraded to 22.3.1)
I pip install nomadnet
I get a fail because rust is not installed.

The whole point of Nomadnet was to use the RPIs in low power/offgrid/isolated situations where they communicate via VHF (its the only way through the forest) so....I am out. I will check back periodically to see if there is any news. Thanks for all you do!

**Reply by markqvist** [reply]
On 32-bit Raspberry Pi OS, there is unfortunately no updated `cryptography` package in the `pip` repositories. As a last effort, you might make it work with:

`sudo apt install python-cryptography`

I am not sure what version is in that, it might work.

The whole dependency packaging circus for the cryptography package on the Pi OS is something that has been bothering me to the point of actually shipping my own binary crypto primitives for RNS ;)

That being said, it *is* usually just a matter of a few commands to fix it, but it often changes, and I have not kept up with the 32-bit releases lately. I'll look into it soon though, and update the install guide.

**Comment by Dualfuel671** [comment]
Ok, I tried sudo apt install python-cryptography to no avail. 

So.....I very frustrated with this.
Here is what I want:
include a list of computers that will actually work with this software.
Include a list of operating systems that actually will work with this software.
In your example you say a RPi could be used....be specific...do you mean a RPI4? its 64bit.
Be specific about the operating system...linux yes, but what linux? obviously not Rasbian, Raspberry Pi OS, or Debian. Will linux mint work?
I bought a brand new computer to try with this, but its Windows 11 and I ended up giving it to my wife so I don't void the warranty. 
What computer (preferably a laptop with ssd, so it can operate in a subzero radio shack) do you recommend? I would love to give this a go, but apparently I am too ignorant about the subtleties.

**Reply by fab1an2** [reply]
Orange Pi Zero  Armbian 22.11.4 Jammy with Linux 5.15.89-sunxi
instaled rns I have trouble with configuration but probably all work ok.

**Reply by ve3jec** [reply]
I have it running on dual VM's using Mint Xfce 21.1 64bit

**Comment by Dualfuel671** [comment]
I put nomad net on two Ubuntu 20.04 machines. I am reading the manual. I have both machines on the same wifi network.
Do I need to include a TCPinterface entry in the /.nomadnet/config file or the /. reticulum/config file?
I am not sure how to contact the other computer. There are two different 16bit numbers.

**Comment by Dualfuel671** [comment]
Ok, finally some progress. By placing the Zurich RNS in the nomadnet config file, I got connected to a peer. Not sure if its a bot, or what.
I am reading a printed version of the manual. I have not come across, or didnt recognize the part about whether messages linger on a bbs, like packet, or if I close Nomadnet, the messages vanish. What is the check mark to the left of a message in the conversations screen?


So if you have two machines or instances of nomadnet on the same wifi network, do you have to put the hash/identity number in the config file? I tried contacting the 16 address directly and got the no keys error. Do I have to create some sort of type = TCPClientInterface  for connecting between peers on the same wifi network, in the config file?

I had the library print out "Reticulum Network Stack, Release 0.4.2 beta" ,it jammed the printer. Then the IT guy read the title...I thought I was going to be led into a little room and questioned...I started prattling about ax.25 and 1200baud rates, and his eyes glazed over...good thing!

**Reply by markqvist** [reply]
> I had the library print out "Reticulum Network Stack, Release 0.4.2 beta" ,it jammed the printer. Then the IT guy read the title...I thought I was going to be led into a little room and questioned...I started prattling about ax.25 and 1200baud rates, and his eyes glazed over...good thing

Haha, oh my, this is hilarious :D

**Comment by Dualfuel671** [comment]
Ok, good news! Reticulum simply does not work right out of the box on 32bit raspberry pi os (or 32bit Linux OS). It DOES work out of the box, on the 64bit Raspberry Pi OS. RPi 3b+, RPi 4, and Pi0W2, Pi 400, and RPi A2 are all 64bit processors.

**Comment by Dualfuel671** [comment]
Big day today....I connected two peers directly and sent messages back and forth. 
When I first had both peers loaded with Nomadnet, I copied the ADDRESS (not the identity) from the log. I was confused about which one to copy into the new conversations file. After the address is typed in, I get an error saying I cannot connect because the keys havent been sent. There was some advice about querying the peer. This is where it got surreal. I could not press Control-e to query the peer...UNTIL, I clicked on the little white underscore mark at the bottom of the conversations page. Then the menu options bar at the bottom of the page, changed over and Control-e was an option.
I don't know how exactly I got the two peers talking yet. I sent the address to peer A and nothing. I sent the address of peer A to peer B and it magically connected. Inbetween that time I sent the query and keys several times, so perhaps that had something to do with it.

In other news, one peer is a Raspberry Pi 3b+ with the 64bit Raspbery Pi OS. So that is the trick...make sure and have a 64bit processor to painlessly install Reticulum.

I really do not know how to set up an ESP32 or ESP8226 tho...are there any places where its mentioned in the manual or discussions? 
When I get to a certain point, I will start a new thread with detailed instructions. There are quirks that need to be mentioned to prevent the ignorant from too much head scratching.

**Reply by markqvist** [reply]
It's really helpful that you are documenting all this. I'll be making a few video tutorials for everything soon, and your input is really valuable in knowing what kind of stuff to focus on and how to put it together. I'm sorry I haven't been able to offer you much advice along the way, but I have just had a huge backlog of messages and requests to deal with. Thanks for your persistence :)

**Comment by Dualfuel671** [comment]
about esp32
section 5.1.3 pg 45 of the manual needs to be revised. I installed rnodeconf and ran auto install and got a error telling me my software was out dated and to uninstall it....this all from the manual dated 24 Nov 2022. Its Dec 13 2022....talk about a quick update cycle!
so when I run rnodeconf --autoinstall I get a list of serial ports. In my case, it was ttyUSB0. I plugged in my esp32 and checked dmesg, to confirm which port it was mounted at. When the rnodeconf program asked me which port, I typed in the full address /dev/ttyUSB0 and I got an error saying that no such port existed and the program exited.
dualfuel@bruce:~$ rnodeconf --autoinstall

Hello!

This guide will help you install the RNode firmware on supported
and homebrew devices. Please connect the device you wish to set
up now. Hit enter when it is connected.

Detected serial ports:
  [1] /dev/ttyUSB0 (CP2102 USB to UART Bridge Controller, 0001)

What serial port is your device connected to? [1] /dev/ttyUSB0 (CP2102 USB to UART Bridge Controller, 0001)
That port does not exist, exiting now.
dualfuel@bruce:~$ 
"Bwahahahaha" laughs the man behind the curtain!

**Reply by markqvist** [reply]
>  I installed rnodeconf and ran auto install and got a error telling me my software was out dated and to uninstall it....this all from the manual dated 24 Nov 2022. Its Dec 13 2022....talk about a quick update cycle!

Yes, a lot has happened lately :) That's right, that section needs updating. Thanks for noting it!

>  I typed in the full address /dev/ttyUSB0 and I got an error saying that no such port existed and the program exited.

Just type in the number of the port, and hit enter. In your case, it looks like you need to use port number !, so just type a "1" and hit enter.

**Comment by Dualfuel671** [comment]
ESP32-WROOM-32U with good data-USB cable. I loaded the ESP32 deepsleep example from the Arduino IDE.  I got good output from the serial monitor. I know that the laptop/console, the ESP32, AND the cable are functioning. I closed the Arduino IDE. Nothing should be competing with rnodeconf. 
this the result,
dualfuel@bruce:~$ rnodeconf --autoinstall

Hello!

This guide will help you install the RNode firmware on supported
and homebrew devices. Please connect the device you wish to set
up now. Hit enter when it is connected.

Detected serial ports:
  [1] /dev/ttyUSB0 (CP2102 USB to UART Bridge Controller, 0001)

What serial port is your device connected to? ttyUSB0
That port does not exist, exiting now.
dualfuel@bruce:~$ 

So I better quit for today and think about it. I wonder why the rnodeconf throws that error...?

**Comment by Dualfuel671** [comment]
So I have been taking a class in how to write JAVA, inorder to help the highschool robotics team. I don't know much yet, but it has got me thinking more like a coder then a victim. I went in to your rnodeconf.py file and found the string "What serial port is your device connected to? " then I looked above it...on line 1078 ports_list =int(input()) kind of gave me a clue....The input for what port you want to use is looking for an integer and I kept stubbornly typing in a string, hence getting tossed out. I am not wrong. The damn name of the port in Linux IS ttyUSB0, and saying anything else gets you to Nergensdorf pretty fast. Sooooo, I would add "select one of the bracketed numbers" to the "which port" string so the user has some idea what to select.
please understand, if we are coming to Reticulum, and know what it is...we know just enough to be dangerous.

**Reply by Dualfuel671** [reply]
[11:29:40] EEPROM Bootstrapping successful!

RNode Firmware autoinstallation complete!

To use your device with Reticulum, read the documetation at:

https://markqvist.github.io/Reticulum/manual/gettingstartedfast.html

Thank you for using this utility! Please help the project by
contributing code and reporting bugs, or by donating!

Your contributions and donations directly further the realisation
of truly open, free and resilient communications systems.

  Ethereum : 0x81F7B979fEa6134bA9FD5c701b3501A2e61E897a
  Bitcoin  : 3CPmacGm34qYvR6XWLVEJmi2aNe3PZqUuq
  Monero   : 87HcDx6jRSkMQ9nPRd5K9hGGpZLn2s7vWETjMaVM5KfV4TD36NcYa8J8WSxhTSvBzzFpqDwp2fg5GX2moZ7VAP9QMZCZGET
  Ko-Fi    : https://ko-fi.com/markqvist

  Info     : https://unsigned.io/
  Code     : https://github.com/markqvist


!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!Well alrighty then, I have validated myself as a human being again...at least for today.

I know its not hip, but you should have a link to Paypal....for the GenXers who lost all their Bitcoin by trading it for FTT!

**Reply by markqvist** [reply]
Ah, you figured it out yourself ;) Just started anwering above before I had read this.

> !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!Well alrighty then, I have validated myself as a human being again...at least for today.
> I know its not hip, but you should have a link to Paypal....for the GenXers who lost all their Bitcoin by trading it for FTT!

Hahaha, I guess you are right! I'll add it to the to-do list :)

Really nice to hear you succeeded in getting it flashed! Again, thanks for these notes, it's really useful for me, and I suspect also other people who will end up here searching for answers when they run into trouble.

**Comment by Dualfuel671** [comment]
Ok, and I am back from a couple of days of trying to set up my network.
I rounded up a vhf radio, a Kantronics TNC, and a RPi 3b+.
I thought I could run NOMADNET without a Desktop. I have already sucessfully installed Raspberry Pi OS 64bit with the desktop. I had sucessfully installed nomadnet on the 64bit Desktop Raspberry Pi OS. I had an issue with the RPi OS 64bit. The console was printing bigger then the screen edges. A diffing of the 32bit /boot/config.txt versus the 64bit /boot/config.txt showed that I needed to change disable_overscan=0 
also, 
overscan_left = 30
overscan_right = 30
overscan_top = 30
overscan_bottom = 30

This works with both the desktop and terminal versions. 
with the terminal version, I had to change the terminal font size...
Sudo dpkg-reconfigure console-setup

Just run through the menus and pick what you want.

Then idiocy struck.....even though Nomadnet is pretty basic, it needs a rudimentary GUI. I found it could not work without the mouse. I am pooped for the day but am pleased with the progress so far.

**Reply by markqvist** [reply]
It should definitely work in the console, completely without a mouse, since that is how I use it most of the time myself ;) What's going wrong exactly?

**Reply by Dualfuel671** [reply]
I got nomadnet loaded. It took three tries, but I think it was a bad internet connection.
Once nomadnet started, I didn't find the menu on the bottom of screen. I am not sure how to navigate. I randomly pushed buttons until I got the log screen.

**Comment by Dualfuel671** [comment]
![16712336087694305610285419220892](https://user-images.githubusercontent.com/119222372/208207883-9225a6c2-1f63-4eab-ba33-45350587e61f.jpg)
This a phone photo of the screen.

**Reply by markqvist** [reply]
Aha! That is strange. Thanks for sharing that photo. You are right, for some odd reason, the menus are not being rendered. I'll look into it. This was on headless raspi os 64 bit, right?

**Reply by Dualfuel671** [reply]
Yes, with rpi3b+

**Reply by markqvist** [reply]
Thanks, will have a look at it when I get the chance. As a stopgap you can access the desktopless nomadnet nodes over SSH if that helps.

**Comment by Dualfuel671** [comment]
It would be great if you could point me in the right direction to learn how to use nomadnet with the ESP32/ESP8266. I can flash them, but I don't have any idea how to implement them.

**Reply by markqvist** [reply]
First, flash a supported ESP32 board using the `rnodeconf` program. This program is included with RNS. It's easiest to use the built-in auto-install guide. Just run this command:

`rnodeconf -a`

That should guide you through the flashing process. Once the flashing and bootstrapping completes, the device is now functionally an [RNode](https://unsigned.io/rnode), a general purpose data-radio module, well suited for, amongst other things, use with Reticulum, and therefore also NomadNet.

You then just add an [RNode Interface](https://markqvist.github.io/Reticulum/manual/interfaces.html#rnode-lora-interface) to your Reticulum config file, and NomadNet can use the device for communication.

There is even more info, and links to more in-depth guides, in the [Adding Radio Interfaces](https://markqvist.github.io/Reticulum/manual/gettingstartedfast.html#adding-radio-interfaces) section of the manual.

**Comment by Dualfuel671** [comment]
![VID_20221218_190736751_exported_9476](https://user-images.githubusercontent.com/119222372/208326949-7b856d2f-a750-4c5a-89e9-4802415bf6fd.jpg)

OK, I put in an 8hour day on this.
 1. I assembled a RPi3b+, Kantronics KPC3, and an Alinco data radio. 
 2.  I tested everything using PUtty and the TNC in terminal INTFACE. The radio, TNC, and RPi worked as they should. This proved the cabling and components were going to work.
 3. I set the TNC to INTFACE KISS. I created an network object ax0, using kissattach. I ran a sudo axlisten -a in one terminal and ran sudo axcall in another terminal.
 4. I noted that the instance of axlisten was printing the axcall sending.
 5. I didn't see any transmission light on the Kantronics. Nor did I see the radio transmitting.
 6. I screwed around for hours reading and trying various things. 
 7. I determined that is is something I am missing in the Linux ax25 commands.
 8. I gave up and loaded a Raspberry Pi OS 64bit SD card and installed Nomadnet.
 9. I simply cut and pasted the KISS interface example from the GitHub document to the reticulum config file using desktop/text editor. This was easy because the cut and paste function works in both instances.
 10. I tried it. I used 1234567890abcdef1234567890abcdef as the address I wanted to contact. I got the usual error about not having keys. I pressed control-e and query for keys. Nomadnet queried but the TNC didn't light up.
 11. I went back and looked at the interface file and changed everything to match the PUtty configuration. Namely, a Kantronics KPC3 has a maximum baud rate between the console and the TNC of 9600.
 12. I also uncommented the bit about sending my call sign in ax.25. I also set the id interval to 10seconds.
 13. I saved the file. I rebooted Nomadnet. I repeated the query to the fake address and got the red light on the TNC. 
 14. Nomadnet worked immediately when I had the correct settings, much easier then Linux ax.25tools and apps.
 15. 
![IMG_20221218_085203153](https://user-images.githubusercontent.com/119222372/208328319-ecec04b8-ac15-420b-8168-599f298a7785.jpg)

**Reply by markqvist** [reply]
Great work! This is awesome to hear! And very interesting for me to hear that it works well on Kantronics TNCs out of the box (as long as you set the correct baudrate, of course). That was of course the intention, but you never know if there's some random quirk ;)

Regarding the Linux ax.25 tools, they seem to have fallen a bit into disrepair in the newer distros, with some dropping kernel ax.25 support altogether, and some having bugs/incompatibilities that prevents stuff from working as expected.

You could probably do with a longer ID interval than 10 seconds as well, it's pretty frequent. On the other hand, the ID transmission is very short, so it doesn't really use that much bandwidth, but still, better save what you can. It reminds me though, that I could actually improve the ID beacon logic a bit, and make it work like it does in the `tncattach` program, which is a bit more intelligent.

> Nomadnet worked immediately when I had the correct settings, much easier then Linux ax.25tools and apps.

Really good to hear, this was also the goal :) That, at least when you know the general idea of how to set everything up with Reticulum, getting apps running and talking to the TNC and other hardware should be a lot simpler and easier. The great thing is also, that now you have it working, all programs on the system that use Reticulum can just effeiciently share the TNC hardware without anything extra on your part.

**Reply by srem1** [reply]
no mobility this stuff ;(

**Comment by Dualfuel671** [comment]
Mark,
I set the ID at such a short interval because I wanted to test the radio. Nomadnet did what it advertised. 
1. I queried for keys - a transmission was sent
2. 10 seconds later another transmission was sent.
3. I left it going for an hour while I wrote that above comments and wrote down my notes. No further transmissions were observed. 
4. Very importantly, the traditional packet station I have, did NOT pick up the query or the ID. There are some underlying important implications that I need to explore.
5. I did NOT listen to the raw transmission on a baofeng, nor did I have an instance of axlisten running to observe the raw interpretation of the nomadnet characters.
Next, I will set a second RPi/TNC/radio up at a distance and start sending things back and forth.

Will nomadnet work over ssh/screen ? I am going to try that so as to save the running back and forth across the snow.
I will also test an MFJ TNC as well.

**Reply by markqvist** [reply]
> I set the ID at such a short interval because I wanted to test the radio. Nomadnet did what it advertised.

Ah, of course, that make sense.

> Very importantly, the traditional packet station I have, did NOT pick up the query or the ID. There are some underlying important implications that I need to explore.

Yes, this is the expected behaviour, and most definitely on purpose. Every packet is sent as pure Reticulum frames, and not encapsulated inside AX.25 packets. Doing so would be quite wasterful in terms of bandwidth, and serve no actual purpose, since all protocol functionality needed is handled by Reticulum itself. It would also be a bit bad practice to have packets passed for processing by any AX.25 implementation that might be listening to the medium, even though they are not actually AX.25 packets.

If your other packet station is expecting AX.25 frames, it will reject any packets from Reticulum (even though the TNC is passing them to the OS AX.25 stack), since they do not carry the usual AX.25 structure or checksums. If you set options on your AX.25 station to monitor and pass all packets, even ones failing checksum checks, you should be able to see them. If you just listen to raw frames from the TNC, you will see the raw packets. The ID beacons are encoded as plain UTF-8 strings, so it's directly readable by anything listening to the TNC port.

> Will nomadnet work over ssh/screen ? I am going to try that so as to save the running back and forth across the snow.
I will also test an MFJ TNC as well.

Yes, absolutely. I use it like this quite often, and mostly have permanent servers running in `screen` or `tmux` sessions, and attach to them when I need to interact with them. So it is also fine over SSH.

**Comment by Dualfuel671** [comment]
Thank you for responding.
Today was short, I managed to get the second radio/TNC/RPi powered up.
It got strange. 
With both radios setup, I sent a message between the two on the router. The auto interface worked fine and the messages passed back and forth.
When I commented out the auto interface in both reticulum config files, I got nothing.
The radio that transmitted last night didn't work today. The new radio didn't work either.
I am very tired. I checked lsusb, and saw the ftdi adapter listed. 
The only move I can think of, is to reset the TNC back into NEWUSER mode and transmit with putty. That would confirm the hardware. Then change to KISS mode, and try reticulum.
I vaguely remember something about this when trying TNCATTACH .

**Reply by markqvist** [reply]
Before doing that, maybe check the output of `rnstatus` to confirm whether traffic is actually being sent to the interface? It should list your modem interface along with traffic stats. That way you can see if Reticulum actually sent something out to the modem.

**Reply by markqvist** [reply]
You're also welcome to post your Reticulum config file. I can have a look and see if anything is up with it.

**Comment by Dualfuel671** [comment]
I did a rnstatus on both RPi. Both were shipping data to the serial port.
I took the screws off the first TNC and made a horrifying discovery. Before I had replaced the cover, previously, I had put the reset jumper back on both reset pins... effectively undoing everything I did two days ago. The horrifying part is that I do not remember doing it.
So I quickly set the TNC to KISS mode, and sent a nomadnet message to the other radio....and the TNC lit right up. I saw the transmit power bar on the radio too. So this part is working again.
Radio number two is not transmitting. The TNC is lighting up, and nomadnet is working but no radio. I realize that's not your problem but....it makes me think about testing the hardware with nomadnet. 
Perhaps you already have a script that sends out a call sign once every second. I was thinking of pressing control-i, having it send, then pressing control-i again to have it stop. That way one could observe lights or listen on the frequency.
Anyhow lemme dig around for a different radio

**Comment by Dualfuel671** [comment]
Dear Mark,
I have hit the limit of what I know. I have started learning Python. I have some questions that I am not sure how to ask.
1. How to combine Rnode with the ESPCAM example in the Arduino IDE?
2. Does everything have to be in Python?
3. Can all the dependencies be downloaded and stored locally? I am thinking of how to flash/load Reticulum without the internet.
4. Can an RNODE be connected to a 32bit computer or a Pentium III and send data across Reticulum, like a WIFI radio connected to a computer with no WIFI adapter does?
5. If one created an [[Packet Radio AX.25 KISS Interface]] could the Reticulum network be transmitted over packet digipeaters? I ask this, because we have many old digipeaters here, up in towers that are inaccessible. They have no computers, just the TNC in "TERMINAL" mode. There may be a window for their use, in an emergency, before their batteries fail. 

Once again, THANK YOU, for this very interesting learning experience.
I presently am trying to configure a TNC->Radio cable for the second KISS vhf radio interface. So I am reading and reading. Merry Christmas!

**Reply by markqvist** [reply]
> How to combine Rnode with the ESPCAM example in the Arduino IDE?

That's a much more complex task than I can just answer here in a few lines ;)

> Does everything have to be in Python?

Not necessarily. You could make small wrapper functions in python for the functionality you want, and call that from other programming languages. Full implementations in C, and bindings to other popular languages such as Rust, JS and others are coming in the future.

> Can all the dependencies be downloaded and stored locally? I am thinking of how to flash/load Reticulum without the internet?

Yes, every RNode now comes with an internal [RNode Bootstrap Console](https://unsigned.io/rnode_bootstrap_console/) that contains all the software needed.

> Can an RNODE be connected to a 32bit computer or a Pentium III and send data across Reticulum, like a WIFI radio connected to a computer with no WIFI adapter does?

Yes.

> If one created an [[Packet Radio AX.25 KISS Interface]] could the Reticulum network be transmitted over packet digipeaters?

Yes, the digipeaters will need to be able to pass packets that are 500 bytes long though (excluding AX.25 headers). Most digipeaters should be able to manage that.

**Comment by ve3jec** [comment]
After reading this whole thread i want to thank you for listing out your experience so far, i don't understand much but appreciate the reading and learning some. 

My main though was to mention the RPi hat i have when i was messing with APRS in 2017, its a TNC-Pi from Coastal Chip Works. In looking for a link it seems they are no longer, and its now a MFJ item and double the over double the price.  Was handy as it made for a compact package when used with the 7" touch screen. 

I ended up with the TNC-Pi, and 2 TNC-X from CCW, when i bought the TNC-X it was Identical to the MFJ units just not assembled. Guess that makes it a $60 premium for assembly on the Pi hat, and $92 premium for the TNC-X. 

Cheers all!
Joe

**Reply by Dualfuel671** [reply]
Joe,
I messed with the CCW TNC-PI for a few years. I still have it. 
I ended up just going to the ham swaps and clubs and buying the old Kantronics KPC3 or the MFJ 1276s, flipping them to KISS mode and using them with the pi. The only PITA is the FTDI to RS232 adapter always hoses you when you type an error. So when I type over the packet network I have to live with spelling errors in the messages.
Otherwise, the old TNCs are cheap and insanely durable. Plus, they can stand alone with their radios, for months without attention. 
With tropospheric ducting, my packet station can be reached from the Grand Marais Minnesota node. KE8DCJ-1 is the BBS.

**Reply by ve3jec** [reply]
Fair enough, when i was hunting a TNC the Kantronics and brethren where all more then a new MFJ, then i happened across the CCW stuff. Never had an issues with the hardware, for what i was doing.  I do wish i kept notes of my messing about, as its been a hot minute now and I'd have to start at the beginning again. Still have the hardware though so that's a plus. 

Ended up seeing a post from Hack A Day, about Reticulum and ended up here following the rabbit hole. 

Cheers

**Comment by Dualfuel671** [comment]
"You could make small wrapper functions in python for the functionality you want, and call that from other programming languages. Full implementations in C, and bindings to other popular languages such as Rust, JS and others are coming in the future."

Mark thanks again for the response! Yes, I asked that question without really understanding what the answer would be shaped like. Since then, I switched over to learning Python (I learn Java at the highschool via VS Code, for the First Robotics Team.) on my personal time. 
So, I have been struggling with the new Arduino IDE(2.0)...and even the old one (1.8.19) so I tried Thonny and Micropython on the ESP32 boards. I swear I spend more time troubleshooting bad USB cables and rigging ftdi adapters so I can use 8266 relay boards, then I ever do learning code!

I blew the U-joint out of my pole setting truck, so cannot set up the VHF towers til spring. I have one last bench test to do with Reticulum over VHF, before all that gear can go to the remote tower shack. 
I am at 47N -88W on an island, so lots of lake effect snow. The sun is coming back though and we are starting to get power again. So, the ESP boards are my main focus right now. ESP32 WROOM 32U, has an external antenna connector so I can get some range out of them.
I will check out the bootstrapping feature.

**Comment by payalspatel** [comment]
Hi guys,

Thanks for all the detailed updates on this forum.
I have been trying to install the RNode firmware on my LORA Hat which is connected to my Raspberry Pi(3b - 64 bit OS) but I am stuck after the rnodeconf -a command with the probing device prompt. Anyone has any experience with running Rnode successfully on LORA hats?

Thanks in advance
